Merging updates to landslides.py

saisiddu opened the pull request:
- updated 'landslides_probability.py'
- added 'source_tracking_algorithm.py' as an utility.

kbarnhart commented on landlab/components/landslides/__init__.py:
```diff
@@ -1,5 +1,5 @@
 
-from .landslide import LandslideProbability
```
Is there a backwards compatibility issue with renaming this?

saisiddu replied:
There should not be. The name of the file 'landslides_probability.py' has been changed from 'landslides.py'. But, there is an .__init__ file in that folder that should take care of this. This should be a low level (internal/away from the user) interface. Please correct me if I am wrong.

kbarnhart commented on landlab/utils/source_tracking_algorithm.py:
```diff
@@ -0,0 +1,189 @@
+        'hsd_upstr' maps each Model Domain (MD) grid node to corresponding
+        contributing upstream HSD ids. 'flow_accum' is an array of the number
+        of upstream contributing nodes.
+    """   
```
Provide more detailed information about what is given in hsd_upstr. "maps" is a generic term and mapping can be stored in many ways, so consider expanding in more detail about what each variables contains.

saisiddu replied:
But this is a python dictionary. And in a Python dictionary, a mapping exists between a key and the corresponding values. Do you still think this needs to be addressed?

saisiddu replied:
I am working on clarifying the documentation though. I think I now understand what you meant about mappings.

saisiddu replied:
I did try to add more information. If this is still unclear, can we do the merge and I will keep working on the documentation for a future commit?

kbarnhart commented on landlab/utils/source_tracking_algorithm.py:
```diff
@@ -0,0 +1,189 @@
+    Parameters:
+    ----------
+    hsd_upstr: dictionary
+        'hsd_upstr' maps each Model Domain (MD) grid node to corresponding
```
As in the prior function, expand the description of what hsd_upstr is.

saisiddu replied:
Done!

kbarnhart commented on landlab/utils/source_tracking_algorithm.py:
```diff
@@ -0,0 +1,189 @@
+    """
+    Counts repeated HSD ids and maps MD nodes with unique upstream HSD ids
+    and corresponding fraction of contribution.
+    
```
A comment that is related to the question above regarding if a landlab grid must be fully contained within a HSD grid.

Here, is the assumption that a landlab grid cell is fully contained within a HSD grid cell? I think this is the case. If so, please state explicitly.

saisiddu replied:
Done!

kbarnhart commented on landlab/utils/source_tracking_algorithm.py:
```diff
@@ -0,0 +1,189 @@
+    Returns:
+    -------
+    (unique_ids, fractions): (dictionary, dictionary)
+        Tuple of data. 'unique_ids' maps each MD node with all upstream HSD
```
Here and elsewhere I recommend defining MD and HSD at the first time used within the function (which should probably be in the docstring). Consider also calling the model domain the Landlab Model Domain (LMD) to reinforce the destination that there are two model domains, the landlab domain and the hyrologic model domain.

saisiddu replied:
Well, we have used MD and HSD for Landslides User Manual and we would like to be consistent with these names. I have tried to emphasize that the grid referred here is a Landlab RasterModelGrid instance though.

kbarnhart replied:
@saisiddu Please add a definition of the MD and HSD the first time they are used (or at least in the component docstring) that includes what is expected of these two grids. 

Otherwise someone might not know that the HDS must be equal in resolution or coarser (I think), etc.

kbarnhart commented on landlab/components/landslides/tests/test_landslide.py:
```diff
@@ -1,5 +1,5 @@
 """
-Unit tests for landlab.components.landslides.landslide
+Unit tests for landlab.components.landslides.landslide_probability
```
Recommend adding a test for each of the distribution options that are specified by groundwater__recharge_distribution.

saisiddu replied:
Will do this for future commit as we are planning to submit the paper tomorrow.

saisiddu replied:
Actually Done! - left out data driven option because, it can get very messy. Other three input methods now have unit tests!

kbarnhart commented on landlab/utils/tests/test_source_tracking_algorithm.py:
```diff
@@ -0,0 +1,57 @@
+from landlab.components import FlowRouter
+from landlab.utils import (track_source,
+                           find_unique_upstream_hsd_ids_and_fractions)
+
```
Recommend adding tests to verify that flow_directions can be specified multiple ways including that you throw the correct errors if flow_directions is misspecified.

saisiddu replied:
Will reserve this for a future commit.

kbarnhart commented on landlab/utils/source_tracking_algorithm.py:
```diff
@@ -0,0 +1,189 @@
+    if flow_directions==None:
+        r = grid.at_node['flow__receiver_node']
+    else:
+        r = flow_directions
```
I think this function assumes that flow__receiver_node already exists as a field, or are the shape of fields at node. 

Recommend the following:
1) First, if flow_directions is None, check that grid.at_node['flow__receiver_node'] exists. If it doesn't then either throw an error OR automatically run a FlowDirector to get the flow__reciever_node variable. D4 is provided by FlowDirectorSteepest and D8 is provided by FlowDirectorD8. 

Second, if flow_directions is provide, check that the provided parameter flow__receiver_node is of the right size (and made of ints that are in the correct range) and throw an error if either of these options are not true.

saisiddu replied:
as mentioned above, we will automatically get a field error if the field is not provided or if the field is  not of the same length as number of nodes. will work on it for future commits.

kbarnhart commented on requirements.txt:
```diff
@@ -12,3 +12,4 @@ six
 setuptools>=18.0
 pyyaml
 xarray
+statsmodels
```
Is statsmodels a stable enough/supported enough library that it is OK to add it as a dependency @mcflugen

mcflugen replied:
I don't know anything about it. However, it is offered with Anaconda (and PyPI) so it's probably pretty stable. How necessary is it @saisiddu? It looks like you just use if for `empirical_distribution` in one location so probably not a big deal include it.

saisiddu replied:
@mcflugen: Are you implying that we should include another method that emulates the empirical_distribution in our component? Or should leave stats_model as a requirement?

kbarnhart commented on landlab/components/landslides/landslide_probability.py:
```diff
@@ -0,0 +1,535 @@
+    stability index (Factor of Safety).
+
+    The driving force for failure is provided by the user in the form of
+    groundwater recharge; 4 options for providing recharge are included.
```
consider changing to "four options for recharge distribution are supported"?

saisiddu replied:
done

kbarnhart commented on landlab/components/landslides/landslide_probability.py:
```diff
@@ -0,0 +1,535 @@
+    groundwater recharge; 4 options for providing recharge are included.
+    The model uses topographic and soils characteristics provided as input
+    by the user in the landslide_driver.
+
```
I'm not sure what the landslide_driver is. Please expand description (here or earlier in the docstring where landslide driver is mentioned). Based on the code it appears that either parameters can be specified as keyword arguments OR parameters can be given in an input file. 

I suspect this is a holdover from when most landlab components were specified with input files.

saisiddu replied:
Answered above.

kbarnhart commented on landlab/components/landslides/landslide_probability.py:
```diff
@@ -0,0 +1,535 @@
+
+    A LandslideProbability calcuation function provides the user with the
+    mean soil relative wetness and probabilty of failure at each node.
+
```
By  a LandslideProbability calculation function, do you mean to say something like:

The main method of the LandslideProbability class is calculate_landslide_probability() which calculates the mean soil relative wetness and probabilty of failure at each node based on a Monte Carlo approach.

kbarnhart replied:
I think adding something about the HSD grid (and assumptions about how it relates spatially to the landlab grid) would be helpful at this point in the code.

saisiddu replied:
Done

saisiddu replied:
We have provided contextual description and referred to the user manual and the paper instead, as the description of these parameters is quite lengthy.

kbarnhart commented on landlab/components/landslides/landslide_probability.py:
```diff
@@ -0,0 +1,535 @@
+                 groundwater__recharge_standard_deviation=None,
+                 groundwater__recharge_HSD_inputs=[],
+                 seed=0, **kwds):
+
```
Perhaps I am mistaken, but some of these parameters are listed below but not in the class parameter definition (e.g. g), and some are listed here, but not the class definition (e.g. seed). Recommend revising for consistency.

saisiddu replied:
nice catch! Done!

kbarnhart replied:
Great! Also, I don't think there are spaces between the docstring and first line of code in PEP8.

kbarnhart commented on landlab/components/landslides/landslide_probability.py:
```diff
@@ -0,0 +1,535 @@
+
+
+    def _calculate_HSD_recharge(self, i):
+        store_Re = np.zeros(self.n)
```
Needs docstring.

saisiddu replied:
Done!

kbarnhart commented on landlab/components/landslides/tests/test_landslide.py:
```diff
@@ -114,3 +111,38 @@ def test_field_initialized_to_zero():
            LS_prob.grid.number_of_nodes))
+
+
+def test_calculate_landslide_probability():
```
Docstring on all unit tests.

mcflugen commented on landlab/utils/source_tracking_algorithm.py:
```diff
@@ -0,0 +1,189 @@
+    receiver_nodes: array_like
+        A numpy.array of shape=[grid.number_of_nodes]. Please note that this
+        array gives the receiver nodes only for the core nodes. For non-core
+        nodes, a zero is used.  
```
For this sort of description include the type and shape of the array on the rist line,
```
receiver_nodes : ndarray of int, shape (n_nodes, )
```

mcflugen commented on landlab/utils/source_tracking_algorithm.py:
```diff
@@ -0,0 +1,189 @@
+    """
+    Convert Arc flow_directions file to be represented in node ids
+    gives receiver of each node starting at right and going clockwise.
+    
```
Always include a one-line description on the first line of the docstring, a blank line, and then the longer description.

saisiddu replied:
Done!

mcflugen commented on landlab/utils/source_tracking_algorithm.py:
```diff
@@ -0,0 +1,189 @@
+    hsd_ids: array_like
+        numpy.array of shape=[grid.number_of_nodes]; array that maps the
+        nodes of the grid to, possibly coarser, Hydrologic Source Domain (HSD)
+        grid ids.
```
As before,
```
hsd_ids : ndarray of int, shape (n_nodes, )
```

saisiddu replied:
Done!

mcflugen commented on landlab/utils/source_tracking_algorithm.py:
```diff
@@ -0,0 +1,189 @@
+        of upstream contributing nodes.
+    """   
+    if flow_directions==None:
+        r = grid.at_node['flow__receiver_node']
```
It doesn't really matter but better to use
```python
if flow_directions is None:
```

saisiddu replied:
Done!

saisiddu replied:
We will see a field error if no flow directions are provided. Hence, leaving it for now. 
@kbarnhart: I will work on adding a call to D8 flow router in a future commit.

mcflugen commented on landlab/utils/source_tracking_algorithm.py:
```diff
@@ -0,0 +1,189 @@
+            # stream segment.
+            if switch_i:
+                switch_i = False
+    return (hsd_upstr, flow_accum)
```
Is this slow? It looks like it might need some optimization. Maybe not now but perhaps something to address in the future?

saisiddu replied:
Sure!

SiccarPoint commented on landlab/components/landslides/landslide_probability.py:
```diff
@@ -0,0 +1,535 @@
+        list of 3 dictionaries in order - HSD_dict {Hydroligic Source
+        Domain (HSD) keys: recharge numpy array values}, {node IDs keys:
+        list of HSD_Id values}, HSD_fractions {node IDS keys: list of
+        HSD fractions values} (none)
```
@mcflugen : wait wait wait. I thought the style was Standard Name style, with the double underscore perfectly fine? So this would be `groundwater__recharge_hsd_inputs`? Vast numbers of our other names look like this (e.g., https://github.com/landlab/landlab/wiki/Standard-names, and Table S1 in Hobley et al)

mcflugen replied:
I'm just talking about variable names not the the strings we use for field names. So,
```python
grid.at_node['groundwater__recharge_hsd_inputs']
```
is correct but
```python
grid._groundwater__recharge_hsd_inputs
```
is not (or as a keyword). Do you not agree?

mcflugen replied:
In addition, just as a matter of style, we should probably just use these really long names for the names of fields. It becomes a bit unwieldy when variables and keywords are so long.

SiccarPoint replied:
Sorry, I misread this. This is all good.

kbarnhart commented:
@saisiddu @RondaStrauch 
in our land lab call today we've been discussing how to deal with a potential memory leak. Here is some code that goes through what I've done to profile to total amount of memory used by a process. 

A note, when I run this on my computer, I get the total amount of memory used at that time, which includes what ever my computer needs to run background processes, run Chrome, etc, etc. So I would use this sort of profiling on a unix box/HPCC/etc that is just running your process. 
 
```python
import resource
usage_file = open('usage.txt', 'w')

# do some stuff...

# get usage information
usage = resource.getrusage(resource.RUSAGE_SELF)

# print out some information of usage. 
for name, desc in [
    ('ru_utime', 'User time'),
    ('ru_stime', 'System time'),
    ('ru_maxrss', 'Max. Resident Set Size'),
    ('ru_ixrss', 'Shared Memory Size'),
    ('ru_idrss', 'Unshared Memory Size'),
    ('ru_isrss', 'Stack Size'),
    ('ru_inblock', 'Block inputs'),
    ('ru_oublock', 'Block outputs'),
    ]:
    usage_file.write('%-25s (%-10s) = %s \n'%(desc, name, getattr(usage, name)))

# In my code, I do some more things, then re-write the usage information. You can modify as you need. 

# Close usage file. 
usage_file.close()
```

I've also found this useful for profiling single variables. 

https://stackoverflow.com/questions/1331471/in-memory-size-of-a-python-structure

saisiddu commented on landlab/components/landslides/landslide_probability.py:
```diff
@@ -0,0 +1,535 @@
+
+Relative wetness and factor-of-safety are based on the infinite slope
+stability model driven by topographic and soils inputs and recharge provided
+by user in a "landslide_driver" file. In addition, the component simulates
```
We have included a reference manual and the paper citation as references. This should take care of users' ambiguity regarding this matter.

saisiddu commented on landlab/components/landslides/landslide_probability.py:
```diff
@@ -0,0 +1,535 @@
+    Instantiate the 'LandslideProbability' component to work on this grid,
+    and run it.
+    
+    >>> LS_prob = LandslideProbability(grid)
```
Done

saisiddu replied:
So, we have removed all capital letters used as parameters in doc tests that has an interface with the user. We have made sure that all parameters/fields have lower case letters and followed our coding conventions. However, we are reserving our rights to use capital letters to name the variables that may not be in direct interface with the user. Please let us know if this is an issue because it would be a huge time sink and challenging (to come up with names that define these variables logically but concisely).

saisiddu commented on landlab/components/landslides/landslide_probability.py:
```diff
@@ -0,0 +1,535 @@
+        # Store grid and parameters and do unit conversions
+        self._grid = grid
+        self.n = int(number_of_iterations)
+        self.g = 9.81
```
Done

saisiddu replied:
We are using 'self.g' instead of 'self._g'. We have been consistently ignoring underscores. Should this be changed? If it is a must, we will invest time for that. If not, we would like to leave the variables, as they are.

saisiddu commented on landlab/components/landslides/landslide_probability.py:
```diff
@@ -0,0 +1,535 @@
+        np.random.seed(seed)
+
+
+    def _interpolate_HSD_dict(self):
```
Done!

kbarnhart replied:
Thanks for adding the docstring. However, here and elsewhere I think you need to add more information about the HSD_dict.

saisiddu commented on landlab/components/landslides/landslide_probability.py:
```diff
@@ -0,0 +1,535 @@
+#!/usr/env/python
+
+"""
+landslide_probability.py:
```
Done!

saisiddu commented on landlab/components/landslides/landslide_probability.py:
```diff
@@ -0,0 +1,535 @@
+
+        for name in self._input_var_names:
+            if name not in self.grid.at_node:
+                self.grid.add_zeros('node', name, units=self._var_units[name])
```
Done!

saisiddu commented on landlab/components/landslides/landslide_probability.py:
```diff
@@ -0,0 +1,535 @@
+        self.Cmax = self.grid.at_node['soil__maximum_total_cohesion'][i]
+        self.phi_mode = self.grid.at_node['soil__internal_friction_angle'][i]
+        self.rho = self.grid.at_node['soil__density'][i]
+        self.hs_mode = self.grid.at_node['soil__thickness'][i]
```
Done!

saisiddu commented on landlab/components/landslides/landslide_probability.py:
```diff
@@ -0,0 +1,535 @@
+        self.prob_fail = -9999*np.ones(
+            self.grid.number_of_nodes, dtype='float')
+        self.landslide__factor_of_safety_distribution = -9999*np.ones(
+            [self.grid.number_of_nodes, self.n], dtype='float')
```
Done!

saisiddu commented on landlab/utils/source_tracking_algorithm.py:
```diff
@@ -0,0 +1,189 @@
+    
+    Returns:
+    -------
+    (hsd_upstr, flow_accum): (dictionary, array_like)
```
I tried. Hopefully the info I added helps.

saisiddu commented on landlab/utils/__init__.py:
```diff
@@ -3,3 +3,7 @@
 #import landlab.utils.count_repeats
 #from landlab.utils.count_repeats import count_repeats
 from landlab.utils.count_repeats import count_repeated_values
+from landlab.utils.source_tracking_algorithm import (
```
Done!

saisiddu commented:
Hi Katy and Eric, can you please look at our commits that address the issues put forth during the code review? Please let us know if there are any concerns/changes that need to be addressed. Thanks!

saisiddu commented:
@kbarnhart and @mcflugen: We have edited the documentation and the code answering most of the concerns you have raised/recommended. We are reserving few for future commits/work. Tests seem to be passing now - especially that we have corrected a sneaky python 2 int to int division. Please review the corrections (if time permits) and merge. We would like to bump the version once this is merged. Thanks a lot! We greatly appreciate your help.

kbarnhart commented on landlab/utils/source_tracking_algorithm.py:
```diff
@@ -0,0 +1,234 @@
+    (grid.nodes.flatten()) as input for hsd_ids.
+    
+    For more information, refer to source_tracking_algorithm_user_manual
+    at ------.
```
I assume you will eventually make this blank reference @RondaStrauch 's paper.

kbarnhart commented on landlab/utils/source_tracking_algorithm.py:
```diff
@@ -0,0 +1,234 @@
+    The algorithm uses brute force method and is time intensive. This
+    function might not work on huge watersheds (we haven't derived the
+    maximum size of the watershed that this function can execute).
+
```
This description is better, but it is still a bit confusing. I've taken a stab at a revision of the first paragraph here (I turned it into two paragraphs):

This algorithm traverses the grid based on information of flow directions at nodes and at every node identifies _all_ the nodes upstream of a given node. The algorithm creates a dictionary with an entry for each node; a node's entry in the dictionary will contain a list with the node_ids of all upstream nodes. Thus this method permits identification of the source area contributing to each and every node in the model grid. This function is different from a standard flow accumulation routine in that it not only calculates the amount of flow at each node, but records the IDs of _all_ upstream nodes. However, similar to a standard flow accumulation routine, it produces an at_node array of the amount of flow passing through the node. It also differs from a standard flow accumulation routing in that it permits the mapping of flow inputs from a coarser grid to to a finer model grid. 

In its present implementation the algorithm has not been optimized for efficient time use. Its methods are brute force and it should be expected to be time intensive. It is not recommended to be run frequently in a modeling exercise. Due to its intensive nature, this algorithm may fail with large watersheds (a present, the development team has not derived a maximum stable watershed size).

saisiddu replied:
Done!

kbarnhart commented on landlab/utils/source_tracking_algorithm.py:
```diff
@@ -0,0 +1,234 @@
+
+MD - Modeling Domain
+HSD - Hydrologic Source Domain
+"""
```
This would be a great place to add text that describe what the MD and HSD are. Defining them as you do here is a good start but not sufficient.

kbarnhart commented on landlab/components/landslides/landslide_probability.py:
```diff
@@ -0,0 +1,628 @@
+
+
+class LandslideProbability(Component):
+    """
```
I think that the first line of the docstring should be one line long and end with a period. 

SO something like

"Landslide probability component using the infinite slope stability model. "

would perhaps be an appropriate first line.

saisiddu replied:
Done!

kbarnhart commented on landlab/components/landslides/landslide_probability.py:
```diff
@@ -0,0 +1,628 @@
+    calculate_landslide_probability(), which calculates the mean soil relative
+    wetness, probability of soil saturation, and probability of failure at
+    each node based on a Monte Carlo simulation.
+
```
Please also describe what is expected of the HSD grid here since otherwise the description of the parameter groundwater__recharge_HSD_inputs is hard to follow.

saisiddu replied:
Hey Katy, we (Ronda, me and Erkan) decided that this is not a good place to explain what 'HSD' or 'MD' is in detail. We came to the conclusion that one should refer to the paper and the user manual (especially to address this issue) if the user is inclined towards using this optional input method. To use this input, the user has to absolutely refer to the references mentioned, to even make sense of the inputs/outputs. 
However, I have made that clear now in the description.

kbarnhart commented on landlab/components/landslides/landslide_probability.py:
```diff
@@ -0,0 +1,628 @@
+            raise ValueError('You must now provide an existing grid!')
+
+    def calculate_factor_of_safety(self, i):
+
```
No space here in PEP8 I think.

saisiddu replied:
Done!

kbarnhart commented on landlab/components/landslides/landslide_probability.py:
```diff
@@ -0,0 +1,628 @@
+        i: int
+            index of core node ID.
+        """
+
```
No space here in PEP8 I think.

saisiddu replied:
Done!

kbarnhart commented on landlab/components/landslides/landslide_probability.py:
```diff
@@ -0,0 +1,628 @@
+            sequence. To create a certain sequence repititively, use the same
+            value as input for seed.
+        """
+
```
No space here in PEP8, I think.

saisiddu replied:
Done!

kbarnhart commented:
@saisiddu I've looked through the modified code and it looks much better. Two overarching comments that I think should be addressed before you merge. They have only to do with docstring documentation and not with the code itself (which is well commented). 

1) Nitpicky comment about PEP8 formatting for docstrings.
some general themes. I've pointed out some places in the code where these are an issue -- but not everywhere:
a) no spaces before or after docstring lines
b) One line docstrings should be a summary line on one line and multi-line docstrings should have a summary line. 
For reference, here is the info about docstring formatting. 
https://www.python.org/dev/peps/pep-0257/

2) Please add more information about the HSD grid in general and the HSD_dict parameter. It is very hard from the existing documentation to understand what is expected about the HSD grid. It might be coarser... but can it be finer? Do the boundaries of the HSD grid cells have to line up with the boundaries of the MG grid cells or can a MG have inputs from 2 HSD cells? Where might one get the HSD grid? 

I suspect you have been trying to keep the documentation general because there are multiple types of sources for this sort of boundary condition information, but providing a specific example to illustrate how this functionality can be used may make it clearer to a user.

saisiddu commented on landlab/components/landslides/landslide_probability.py:
```diff
@@ -0,0 +1,535 @@
+            Domain (HSD) keys: recharge numpy array values}, {node IDs keys:
+            list of HSD_Id values}, HSD_fractions {node IDS keys: list of
+            HSD fractions values} (none)
+        g: float, optional
```
They should be the same for __init__ and the component.

saisiddu commented on landlab/components/landslides/landslide_probability.py:
```diff
@@ -0,0 +1,535 @@
+
+        # Raise an error if somehow someone is using this weird functionality
+        if self._grid is None:
+            raise ValueError('You must now provide an existing grid!')
```
Done!

saisiddu commented on landlab/components/landslides/landslide_probability.py:
```diff
@@ -0,0 +1,535 @@
+        if self._grid is None:
+            raise ValueError('You must now provide an existing grid!')
+
+    def calculate_factor_of_safety(self, i):
```
Yes, it can be. We shall try that in future.

saisiddu commented:
@kbarnhart :
Hi Katy, we have addressed these issues as much as we could. Erkan, Ronda and I are convinced that the information provided, along with the cited references, will serve the purpose of educating the user about how to use the component. If one has to use this component, we feel that the user should definitely refer the paper and the user manual without fail. Since the tests are passing, and if you don't find any other major issues, can you please merge this pull request? Thank you very much for your time and efforts. We greatly appreciate it.

kbarnhart commented:
Thanks for adding the additional bits of documentation @saisiddu! I'm agnostic to where information about HSD grid is, so I think it is totally fine to just point to other documentation (as long as its pointing somewhere useful). 

Fine with me to merge. @mcflugen do you need to look again, or is it OK to merge.

saisiddu commented:
Since Katy is fine to merge and Eric is traveling, I will go ahead and merge this pull request.
